examples/vision/super_resolution_sub_pixel.py

```diff
@@ -0,0 +1,414 @@
+    plt.title(title)
+    # zoom-factor: 2.0, location: upper-left
+    axins = zoomed_inset_axes(ax, 2, loc=2)
+    axins.imshow(img_array[::-1], origin="lower")
+
+    # Specify the limits.
+    x1, x2, y1, y2 = 200, 300, 100, 200
+    # Apply the x-limits.
+    axins.set_xlim(x1, x2)
+    # Apply the y-limits.
+    axins.set_ylim(y1, y2)
+
+    plt.yticks(visible=False)
+    plt.xticks(visible=False)
+
+    # Make the line.
+    mark_inset(ax, axins, loc1=1, loc2=3, fc="none", ec="blue")
+    plt.savefig(str(prefix) + "-" + title + ".png")
+    plt.show()
+
+
+def predict_result(model, img):
+    """Predict the result based on input image and restore the image as RGB."""
+    ycbcr = img.convert("YCbCr")
+    y, cb, cr = ycbcr.split()
+    y = img_to_array(y)
+    y = y.astype("float32") / 255.0
+
+    input = np.expand_dims(y, axis=0)
+    out = model.predict(input)
+
+    out_img_y = out[0]
+    out_img_y *= 255.0
+
+    # Restore the image in RGB color space.
+    out_img_y = out_img_y.clip(0, 255)
+    out_img_y = out_img_y.reshape((np.shape(out_img_y)[0], np.shape(out_img_y)[1]))
+    out_img_y = PIL.Image.fromarray(np.uint8(out_img_y), mode="L")
+    out_img_cb = cb.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img_cr = cr.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img = PIL.Image.merge("YCbCr", (out_img_y, out_img_cb, out_img_cr)).convert(
+        "RGB"
+    )
+
+    return out_img
+
+
+def get_lr_hr_predict(model, upscale_factor, img_path):
+    """Build low-resolution input and high-resolution original
+  image and prediction result to compare."""
+    img = load_img(img_path)
+    img_arr = img_to_array(img)
+    lowres_input = img.resize(
+        (img.size[0] // upscale_factor, img.size[1] // upscale_factor),
+        PIL.Image.BICUBIC,
+    )
+    w = lowres_input.size[0] * upscale_factor
+    h = lowres_input.size[1] * upscale_factor
+
+    lowres_img = lowres_input.resize((w, h))
+    highres_img = img.resize((w, h))
+    out_img = predict_result(model, lowres_input)
+    return lowres_img, highres_img, out_img
+
+
+"""
+### Define callbacks to monitor training
+
+PSNR: We use PSNR to evaluate our model and please check it from
+[here](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio) for more details.
+"""
+
+import math
+
+
+class ESPCNCallback(keras.callbacks.Callback):
+    # Store PSNR value in each epoch.
+    def on_epoch_begin(self, epoch, logs=None):
+        self.psnr = []
+        self.test_img = test_img_paths[0]
+
+    def on_epoch_end(self, epoch, logs=None):
+        print(
+            "End epoch {} of training and loss is {:7.4f} "
+            "and the average PSNR is {:7.4f}".format(
+                epoch + 1, logs["loss"], sum(self.psnr) / len(self.psnr)
+            )
+        )
+        _, _, predict = get_lr_hr_predict(self.model, upscale_factor, self.test_img)
+        plot_results(predict, "epoch-" + str(epoch), "prediction")
+
+    def on_train_batch_end(self, batch, logs=None):
```

Comment:
Printing any metric computed on a single batch is not useful, because it's not accurate.

What you should print is the metric average over the epoch.

For the loss, the progress bar already does this.

For `psnr`, you can use a `keras.metrics.Mean` metric tracker, and do:

- in __init__: `self.psnr_tracker = keras.metrics.Mean('psnr')`
- in on_batch_end: `self.psnr_tracker.update_state(psnr)`
- in on_epoch_end:  `print('Mean PSNR for epoch: %.2f' % (self.psnr_tracker.result(),)` and `self.psnr_tracker.reset_states()`